backend/tournesol/views/polls.py

```diff
@@ -136,6 +146,40 @@ def filter_unsafe(self, queryset, filters):
                 tournesol_score__gt=0,
             )
 
+    def sort_results(self, queryset, filters):
+        """
+        Sort the results
+
+        For full-text searches:
+        Calculate a "search score" based on the criteria scores and
+        the search relevance calculated previously by Postgres.
+        There are many possible ways to calculate it, some concerns are:
+        - Prefer showing high-rated content
+        - A bad but relevant content should have a higher search score
+            than a great but totally irrelevant content
+        - The search score should always be increasing with the relevance and total score
+
+        This lead to the choice of a model of the type:
+            search_score = relevance *
+                           (weights_sum * relevance +
+                           search_score_coef * (total_score + weights_sum))
+
+        With "search_score_coef" an arbitrary, positive value.
+            If set to 0, total_score is ignored.
+            Set it to a higher value to take more into account the total score.
+        """
+        if filters["search"]:
+            weights_sum = self._weights_sum
+            queryset = queryset.annotate(
+                search_score=F("relevance") * (
+                     weights_sum * F("relevance") +
+                     self.search_score_coef * (weights_sum + F("total_score"))
+                )
```

Comment:
> For entities with negative total_score, the search_score may be negative too. As result the most relevant results will be put even lower in the ranking (due to the multiplication by "relevance").

I used `self.search_score_coef * (weights_sum + F("total_score"))` instead of `self.search_score_coef * F("total_score")`. So that, assuming that `weights_sum > 0`  and `F("total_score") <= weights_sum` (if weights_sum is not the maximum value of total_score for the given weights, replace it by this maximum value), the result should be always positive.

> the ranking is still sensitive to the scale of the weights. I would expect a similar ranking when all weights equal 10, and when all weights equal 100. With this implementation larger weights may reduce the importance of "relevance".

I may have misunderstood this. My rationale is that when the weights are multiplied by 10, weights_sum will be multiplied by 10, and F("total_score") will also be multiplied by 10. So, search_score would be exactly 10 times higher.

Reply:
Here is an example of what it looks like (with weights_sum equal to 100 and each curve is for a total_score in range(-100, 101, 20)):

![image](https://user-images.githubusercontent.com/47578089/183506674-99e70c54-b97b-4e2c-b570-0eb9ea80246f.png)

Reply:
My bad, I got confused with my second point. The scaling seems to be ok.  

The mistake is due to the assumption that we always have `F("total_score") <= weights_sum`. This is not the case: with the current implementation of Mehestan, each criteria scores belongs to [-100, 100], and when no explicit weight is defined, a single criterion is considered (`weights_sum` = 1). As for Licchavi scores are not bounded, no maximum score can be defined.

Reply:
So, assuming we can ignore Licchavi, which is to be deprecated, we can normalize the total_score using `MEHESTAN_MAX_SCALED_SCORE`.

The implementation would be
```py
        if filters["search"] and self.poll_from_url.algorithm == ALGORITHM_MEHESTAN:
            max_absolute_score = MEHESTAN_MAX_SCALED_SCORE * self._weights_sum
            normalized_total_score = (
                (F("total_score") + max_absolute_score) / 2 * max_absolute_score
            )
            queryset = queryset.annotate(
                search_score=F("relevance")
                * (F("relevance") + self.search_score_coef * normalized_total_score)
            )
            return queryset.order_by("-search_score", "-pk")
        else:
            return queryset.order_by("-total_score", "-pk")
```

@glerzing It should be equivalent with what you described, except Mehestan scaling is now correctly taken into account. What do you think?

Reply:
It looks good, just don't forget the parentheses: `(F("total_score") + max_absolute_score) / 2 * max_absolute_score` => `(F("total_score") + max_absolute_score) / (2 * max_absolute_score)`

And please give a check that it returns the same search_score values in practice.

By the way, I like your other improvement commits 👍.

Reply:
> It looks good, just don't forget the parentheses

@glerzing Thanks for noticing that ^^. All questions have been addressed I think. Could you have a last look before we merge?

Reply:
When using Django Debug Toolbar, it returns an error on the EXPLAIN ANALYZE:
`2022-08-13 00:21:13.458 UTC [150] ERROR:  syntax error at or near "ARRAY" at character 1278`

Which is about this line:
`ON c.oid = e.search_config_name::regconfig AND c.cfgname IN ARRAY['customized_english']`

Something that solves this issue is to instead make:
`ON c.oid = e.search_config_name::regconfig AND c.cfgname = ANY(%s)`
(and to replace `tuple(search_configs)` by `search_configs`)

Do you approve this change?

Reply:
Ok it seems to be caused by a bug in "django-debug-toolbar", which fails to preserve tuple params: https://github.com/jazzband/django-debug-toolbar/issues/1482

I updated to query to use `= ANY(%s)`, just in case it can help with future debugging.

Reply:
We can always set a higher value for search_score_coef if we consider that the search doesn't take enough into account the total score (I was using this script to visualize it: [2d_model.zip](https://github.com/tournesol-app/tournesol/files/9331759/2d_model.zip) ). 

If you think the current value is ok, I have no more comments and we can push.